src/BigNumbers.sol

```diff
@@ -997,8 +997,8 @@ library BigNumbers {
             // with addition, if we assume that some a is at least equal to some b, then the resulting bit length will
             // be a's bit length or (a's bit length)+1, depending on carry bit.this is cheaper than calling bitLength.
             let msword := mload(add(result,0x20))                             // get most significant word of result
-            // if(msword==1 || msword>>(max_bitlen % 256)==1):
-            if or( eq(msword, 1), eq(shr(mod(max_bitlen,256),msword),1) ) {
+            // if(carry==1 || msword>>(max_bitlen % 256)==1):
+            if or( eq(carry, 1), eq(shr(mod(max_bitlen,256),msword),1) ) {
```

Comment:
That should be fairly simple to illustrate with an example, right ?

Reply:
Say we have two normalized big numbers `a = [a2, a1]` and `b = [b1]`.
`a1 = 0000000000000000000000000000000000000000000000000000000000000001`
`a2 = 0000000000000000000000000000000000000000000000000000000000000001`
`b1 = 0000000000000000000000000000000000000000000000000000000000000001`
The result would be `r = [r2, r1] = [a2, a1 + b1]`.
`r1 = 0000000000000000000000000000000000000000000000000000000000000002`
`r2 = 0000000000000000000000000000000000000000000000000000000000000001`

[Here](https://github.com/firoorg/solidity-BigNumber/pull/18/files#diff-60b5d52a5be7b9c2f66d6c75bdb9ab0949e9e671dff2e9b44fb1988f52a70e62R984) you set `1` to the most significant word if an overflow occures. But in some cases, `msword` equals `1` even without overflow, as shown in the example (`r2`). This is why I suggest using the `carry` instead of `msword` to determine whether to increase the bit length.

Reply:
Hi, thanks for your contribution!
would you mind adding a unit test (`BigNumbers.t.sol`) implementing your example?

Reply:
@riordant Done